packages/python-google-compute-engine/google_compute_engine/metadata_watcher.py

```diff
@@ -43,26 +43,21 @@ def __init__(self, response):
 def RetryOnUnavailable(func):
   """Function decorator to retry on a service unavailable exception."""
 
-  @functools.wraps(func)
-  def Wrapper(*args, **kwargs):
-    while True:
-      try:
-        response = func(*args, **kwargs)
-      except (httpclient.HTTPException, socket.error, urlerror.URLError) as e:
-        time.sleep(5)
-        if (isinstance(e, urlerror.HTTPError)
-            and e.getcode() == httpclient.SERVICE_UNAVAILABLE):
-          continue
-        elif isinstance(e, socket.timeout):
-          continue
-        raise
-      else:
-        if response.getcode() == httpclient.OK:
-          return response
-        else:
-          raise StatusException(response)
-  return Wrapper
-
+  def func_retry_condition(e):
+    if (isinstance(e, urlerror.HTTPError)
+        and e.getcode() == httpclient.SERVICE_UNAVAILABLE):
+      return True
+    elif isinstance(e, socket.timeout):
+      return True
+    return False
+
+  def func_response_on_success(response):
+    if response.getcode() == httpclient.OK:
+      return response
+    raise StatusException(response)
+
+  return network_utils.RetryOnUnavailable(func, func_retry_condition,
```

Comment:
moved the retry main logic to network_utils

Reply:
Moved to service_utils, since the retry logic can be shared

Reply:
You can already use `RetryOnUnavailable` from other files since it is a public function in the metadata watcher library. Some users may already be doing this.

Reply:
I'ts not used by others right now.
I may keep the RetryOnUnavailable in the original place, but the problem is code of "RetryOnUnavailable" won't be exactly the same for different users - they have tiny differences on exception handling, behaviors on success, etc. So I made it as a template and accept func (for example, func_response_on_success) as args.
Please check my latest code structure to see whether it's reasonable or not.